62 flatten unflatten c

afowler opened the pull request:
C flatten/unflatten is ready for testing and review. Minor differences to note (that can be changed), flatten returns a (string,string) tuple (the Python flatten returns labRAD types in the second position) and C unflatten takes this tuple as input (the Python unflatten takes the two tuple members as two separate inputs). This means you can write: unflatten(flatten([1,2,3])), and can also read the output of flatten([1,2,3]).

Look forward to feedback!

pomalley commented on labrad/fasttypes.c:
```diff
@@ -0,0 +1,2364 @@
+PyObject *ft_flatten(PyObject *self, PyObject *args, PyObject *keywds);
+
+// functions to create an internal C representation (COBJ) of supported Python objects
+COBJ *ft_create_cobj_from_pyobj(PyObject *obj);
```
did the indentation get messed up after this line or is github displaying wrong?

afowler replied:
Just grouping functions into levels to indicate which are only called by a
higher function.
On Apr 22, 2015 3:59 PM, "pomalley" <notifications@github.com> wrote:

> In labrad/fasttypes.c
> <https://github.com/martinisgroup/pylabrad/pull/88#discussion_r28923040>:
>
> > +PyTypeObject *ft_gv_numpy_int32 = NULL;
> > +PyTypeObject *ft_gv_numpy_uint32 = NULL;
> > +PyTypeObject *ft_gv_numpy_float64 = NULL;
> > +PyTypeObject *ft_gv_numpy_complex128 = NULL;
> > +PyObject *ft_gv_labrad_types = NULL;
> > +PyObject *ft_gv_labrad_units = NULL;
> > +PyTypeObject *ft_gv_labrad_units_WithUnit = NULL;
> > +PyTypeObject *ft_gv_labrad_units_Value = NULL;
> > +PyTypeObject *ft_gv_labrad_units_Complex = NULL;
> > +PyTypeObject *ft_gv_labrad_units_ValueArray = NULL;
> > +
> > +// C equivalent of labrad.types.flatten, called in similar manner, although type tags must be strings
> > +PyObject *ft_flatten(PyObject *self, PyObject *args, PyObject *keywds);
> > +
> > +// functions to create an internal C representation (COBJ) of supported Python objects
> > +COBJ *ft_create_cobj_from_pyobj(PyObject *obj);
>
> did the indentation get messed up after this line or is github displaying
> wrong?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88/files#r28923040>.
>

pomalley commented on labrad/fasttypes.c:
```diff
@@ -0,0 +1,2364 @@
+    COBJ *cobj, *cobj_type, *cobj_tag;
+    
+    if (!ft_gv_initialized) {
+        if (ft_initialize() == EXCEPTION_RAISED) goto exception;
```
goto!!

afowler replied:
Yup, standard way of handling exceptions in the Python C API.
On Apr 22, 2015 4:00 PM, "pomalley" <notifications@github.com> wrote:

> In labrad/fasttypes.c
> <https://github.com/martinisgroup/pylabrad/pull/88#discussion_r28923066>:
>
> > +// C equivalent of labrad.types.unflatten, called in similar manner, although type tags must be strings
> > +PyObject *ft_unflatten(PyObject *self, PyObject *args);
> > +    PyObject *ft_unflatten_partial_parse(PyObject *o);
> > +    PyObject *ft_unflatten_no_parse(char *s, int *si_ptr, COBJ *cobj_type);
> > +
> > +// currently immature test function
> > +PyObject *ft_test(PyObject *self, PyObject *args);
> > +
> > +PyObject *ft_flatten(PyObject *self, PyObject *args, PyObject *keywds) {
> > +    int i, n;
> > +    PyObject *obj, *types, *endianness, *sobj, *lobj, *tobj, *robj;
> > +    char *kwlist[] = {"obj", "types", "endianness", NULL};
> > +    COBJ *cobj, *cobj_type, *cobj_tag;
> > +
> > +    if (!ft_gv_initialized) {
> > +        if (ft_initialize() == EXCEPTION_RAISED) goto exception;
>
> goto!!
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88/files#r28923066>.
>

ejeffrey commented on labrad/fasttypes.c:
```diff
@@ -0,0 +1,2364 @@
+// functions to create an internal C representation (COBJ) of supported Python objects
+COBJ *ft_create_cobj_from_pyobj(PyObject *obj);
+    COBJ *ft_allocate_memory_for_cobj();
+    COBJ *ft_create_empty_cobj();
```
These should be declared explicitly (void).  () doesn't catch errors at call sites that pass unwanted arguments for compatibility with pre-c89 that didn't have prototypes.  (I will push a commit with this and other changes to fix compiler warnings)

ejeffrey commented on labrad/fasttypes.c:
```diff
@@ -0,0 +1,2364 @@
+#define SUCCESS 1
+#define EXCEPTION_RAISED -1
+#define BIG_ENDIAN ">"
+#define LITTLE_ENDIAN "<"
```
On my system, BIG_ENDIAN and LITTLE_ENDIAN are already preprocessor macros.  I will rename.

ejeffrey commented on labrad/fasttypes.c:
```diff
@@ -0,0 +1,2364 @@
+#define LABRAD_FLOAT_SIZE 8
+
+// supported internal data types
+#define EMPTY 0
```
maybe use an enum here?

ejeffrey commented on labrad/fasttypes.c:
```diff
@@ -0,0 +1,2364 @@
+    return size;
+}
+
+void ft_appropriate_m
```
tc_list needs an initializer, other wise if line 2303 causes an exception, it will be use uninitialized in the exception handler.

afowler replied:
Thanks for the improvements!
On Apr 22, 2015 10:05 PM, "ejeffrey" <notifications@github.com> wrote:

> In labrad/fasttypes.c
> <https://github.com/martinisgroup/pylabrad/pull/88#discussion_r28936730>:
>
> > +                ft_appropriate_memcpy(buf+size, &((double *)cobj->arr)[i], LABRAD_FLOAT_SIZE);
> > +                size += LABRAD_FLOAT_SIZE;
> > +            }
> > +        }
> > +        else if (cobj->subtype == COMPLEX128) {
> > +            for (i=0; i<2*n; i++) {
> > +                ft_appropriate_memcpy(buf+size, &((double *)cobj->arr)[i], LABRAD_FLOAT_SIZE);
> > +                size += LABRAD_FLOAT_SIZE;
> > +            }
> > +        }
> > +    }
> > +
> > +    return size;
> > +}
> > +
> > +void ft_appropriate_m
>
> tc_list needs an initializer, other wise if line 2303 causes an exception,
> it will be use uninitialized in the exception handler.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88/files#r28936730>.
>

ejeffrey commented:
Flattening a list with a specified type tag seems to not work:

In [1]: import labrad.fasttypes as ft
In [2]: ft.flatten([1,2,3])
Out[2]: ('\x00\x00\x00\x03\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x03', '*i')
In [3]: ft.flatten([1,2,3], ['*i'])
*** Error in `/usr/bin/python': free(): invalid pointer: 0x00007fa47fd38c10 ***
Aborted (core dumped)

ejeffrey commented:
So I modified types.py to try calling fasttypes, then call the python implementation only if fasttypes raises a TypeError, then I am just running labrad/test/test_types

ejeffrey commented:
I added test_fasttypes with a few methods, including some that fail already.  We should see if travis can handle building the binary module and having it crash the python interpreter.

afowler commented:
Thanks, I'll look into this later today. Should this work? Looks to me like
you have passed a list of type tags, the first of which is a single
integer, which should fail when applied to a list of integers, but
admittedly more gracefully than it currently does. Could you clarify?

On Wed, Apr 22, 2015 at 11:32 PM, ejeffrey <notifications@github.com> wrote:

> Flattening a list with a specified type tag seems to not work:
>
> In [1]: import labrad.fasttypes as ft
> In [2]: ft.flatten([1,2,3])
> Out[2]:
> ('\x00\x00\x00\x03\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x03', '
> *i') In [3]: ft.flatten([1,2,3], ['*i'])
> *** Error in `/usr/bin/python': free(): invalid pointer:
> 0x00007fa47fd38c10 ***
> Aborted (core dumped)
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88#issuecomment-95460013>.
>

afowler commented:
Fasttypes raises more than just typeerror exceptions, also syntax and
memory exceptions. It returns a NULL pointer when it fails, and raises what
it considers the appropriate exception.

On Wed, Apr 22, 2015 at 11:33 PM, ejeffrey <notifications@github.com> wrote:

> So I modified types.py to try calling fasttypes, then call the python
> implementation only if fasttypes raises a TypeError, then I am just running
> labrad/test/test_types
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88#issuecomment-95460879>.
>

ejeffrey commented:
I passed in a list of tags with one element '*i' which should be a list of
integers, so in this case the typetag should be correct.

Separately, we do need flatten to fail with an exception when none of the
listed typetags are work for the data provided.

Evan

On Thu, Apr 23, 2015 at 12:37 AM, afowler <notifications@github.com> wrote:

> Thanks, I'll look into this later today. Should this work? Looks to me like
> you have passed a list of type tags, the first of which is a single
> integer, which should fail when applied to a list of integers, but
> admittedly more gracefully than it currently does. Could you clarify?
>
> On Wed, Apr 22, 2015 at 11:32 PM, ejeffrey <notifications@github.com>
> wrote:
>
> > Flattening a list with a specified type tag seems to not work:
> >
> > In [1]: import labrad.fasttypes as ft
> > In [2]: ft.flatten([1,2,3])
> > Out[2]:
> > ('\x00\x00\x00\x03\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x03', '
> > *i') In [3]: ft.flatten([1,2,3], ['*i'])
> > *** Error in `/usr/bin/python': free(): invalid pointer:
> > 0x00007fa47fd38c10 ***
> > Aborted (core dumped)
> >
> > --
> > Reply to this email directly or view it on GitHub
> > <https://github.com/martinisgroup/pylabrad/pull/88#issuecomment-95460013
> >.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88#issuecomment-95478106>.
>

ejeffrey commented:
==12672== Conditional jump or move depends on uninitialised value(s)
==12672==    at 0xD04CEC9: ft_free_cobj (fasttypes.c:674)
==12672==    by 0xD0508E1: ft_flatten (fasttypes.c:238)
==12672==    by 0x52C6D4: PyEval_EvalFrameEx (ceval.c:4020)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==    by 0x52CF31: PyEval_EvalFrameEx (ceval.c:4106)
==12672==    by 0x52CF31: PyEval_EvalFrameEx (ceval.c:4106)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==    by 0x55C593: PyEval_EvalCodeEx (ceval.c:3252)

ejeffrey commented:
==12672== Invalid free() / delete / delete[] / realloc()
==12672==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12672==    by 0xD04CED3: ft_free_cobj (fasttypes.c:674)
==12672==    by 0xD0508E1: ft_flatten (fasttypes.c:238)
==12672==    by 0x52C6D4: PyEval_EvalFrameEx (ceval.c:4020)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==    by 0x52CF31: PyEval_EvalFrameEx (ceval.c:4106)
==12672==    by 0x52CF31: PyEval_EvalFrameEx (ceval.c:4106)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==  Address 0xb5552d0 is not stack'd, malloc'd or (recently) free'd

ejeffrey commented:
==12672== 4 bytes in 1 blocks are definitely lost in loss record 8 of 1,439
==12672==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12672==    by 0xD04E533: ft_create_cobj_type (fasttypes.c:702)
==12672==    by 0xD04E33B: ft_upgrade_first_cobj_to_encompass_second (fasttypes.c:1017)
==12672==    by 0xD0507B2: ft_flatten (fasttypes.c:227)
==12672==    by 0x52C6D4: PyEval_EvalFrameEx (ceval.c:4020)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==    by 0x52CF31: PyEval_EvalFrameEx (ceval.c:4106)
==12672==    by 0x52CF31: PyEval_EvalFrameEx (ceval.c:4106)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)

ejeffrey commented:
==12672== 100 (96 direct, 4 indirect) bytes in 1 blocks are definitely lost in loss record 677 of 1,439
==12672==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12672==    by 0xD04CDED: ft_create_cobj (fasttypes.c:636)
==12672==    by 0xD04EA3C: ft_create_cobj_tag_from_char_buf (fasttypes.c:1481)
==12672==    by 0xD04EF22: ft_create_cobj_tag_from_pystring (fasttypes.c:1322)
==12672==    by 0xD050FEB: ft_unflatten_partial_parse (fasttypes.c:2141)
==12672==    by 0xD05108C: ft_unflatten (fasttypes.c:2119)
==12672==    by 0x52C6D4: PyEval_EvalFrameEx (ceval.c:4020)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==    by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
==12672==    by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
==12672==    by 0x52CF31: PyEval_EvalFrameEx (ceval.c:4106)

afowler commented:
Okay, ready for another run through valgrind, should be better now. Please
let me know if issues remain.

On Thu, Apr 23, 2015 at 3:23 PM, ejeffrey <notifications@github.com> wrote:

> ==12672== 100 (96 direct, 4 indirect) bytes in 1 blocks are definitely
> lost in loss record 677 of 1,439
> ==12672== at 0x4C2AB80: malloc (in
> /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
> ==12672== by 0xD04CDED: ft_create_cobj (fasttypes.c:636)
> ==12672== by 0xD04EA3C: ft_create_cobj_tag_from_char_buf (fasttypes.c:1481)
> ==12672== by 0xD04EF22: ft_create_cobj_tag_from_pystring (fasttypes.c:1322)
> ==12672== by 0xD050FEB: ft_unflatten_partial_parse (fasttypes.c:2141)
> ==12672== by 0xD05108C: ft_unflatten (fasttypes.c:2119)
> ==12672== by 0x52C6D4: PyEval_EvalFrameEx (ceval.c:4020)
> ==12672== by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
> ==12672== by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
> ==12672== by 0x56D0A9: function_call.15337.3181 (ceval.c:3252)
> ==12672== by 0x52E1E5: PyEval_EvalFrameEx (abstract.c:2529)
> ==12672== by 0x52CF31: PyEval_EvalFrameEx (ceval.c:4106)
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88#issuecomment-95734773>.
>

ejeffrey commented:
@afowler are you able to run the test_fasttypes in a reasonable fashion?  I have added a bunch of tests that fail, if I just push them can you get the test report.

afowler commented:
Thanks Evan, a number of bugs now fixed. All tests passed bar the following
two:

>>> ft.flatten(2**31-1)[1]
'i'
>>> str(T.flatten(2**31-1)[1])
'w'
>>>
labrad.types unnecessarily uses w here

>>> ft.flatten(-2**31)[1]
'i'
>>> T.flatten(-2**31)[1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "c:\git\pylabrad\labrad\types.py", line 326, in flatten
    s, t = _flatten_to(obj, types, endianness)
  File "c:\git\pylabrad\labrad\types.py", line 342, in _flatten_to
    s, t = t.__flatten__(obj, endianness)
  File "c:\git\pylabrad\labrad\types.py", line 590, in __flatten__
    raise ValueError("out of range for type w: {0}".format(n))
ValueError: out of range for type w: -2147483648
>>>
labrad.types fails on this one

On Thu, Apr 23, 2015 at 4:42 PM, ejeffrey <notifications@github.com> wrote:

> @afowler <https://github.com/afowler> are you able to run the
> test_fasttypes in a reasonable fashion? I have added a bunch of tests that
> fail, if I just push them can you get the test report.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88#issuecomment-95749832>.
>

maffoo commented on labrad/fasttypes.c:
```diff
@@ -0,0 +1,2521 @@
+    }
+}
+
+PyObject *ft_flatten_cobj(COBJ *cobj, COBJ *co
```
this assumes that the number of dimensions is < 10, produces invalid type tags otherwise:
```python
>>> a = np.array([1.0]).reshape([1]*10)
>>> T.flatten(a)
('\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x01?\xf0\x00\x00\x00\x00\x00\x00',
 LRList(LRValue(''), depth=10))

>>> ft.flatten(a)
('\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00',
 '*:v[]')
```
Also, note that the dimensions are not flattened correctly and the data does not appear at all. This is happening for me with nd-arrays of 2 or higher dimension on Mac with 64-bit python; none of them flatten properly. Seems like this is probably due to the `PyArrayObject` here using `int *` for the dimensions, which is not correct on 64-bit.

maffoo replied:
The size of this buffer is never checked, which seems really dangerous for unflattening data coming from the network. We should probably 'traverse' the buffer first to verify that it has the right length and contains valid data of the specified type.

afowler replied:
Thanks Matthew, I'll get to work on these.

Regards,
Austin.

On Wed, May 6, 2015 at 6:16 PM, Matthew Neeley <notifications@github.com>
wrote:

> In labrad/fasttypes.c
> <https://github.com/martinisgroup/pylabrad/pull/88#discussion_r29817931>:
>
> > +        case NDARRAY:
> > +            printf("NDARRAY");
> > +            break;
> > +        case NDARRAY_SU:
> > +            printf("NDARRAY_SU");
> > +            break;
> > +        case TIME:
> > +            printf("TIME");
> > +            break;
> > +        case ERROR:
> > +            printf("ERROR");
> > +            break;
> > +    }
> > +}
> > +
> > +PyObject *ft_flatten_cobj(COBJ *cobj, COBJ *co
>
> The size of this buffer is never checked, which seems really dangerous for
> unflattening data coming from the network. We should probably 'traverse'
> the buffer first to verify that it has the right length and contains valid
> data of the specified type.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88/files#r29817931>.
>

maffoo commented:
I ran into a problem unflattening a 2d-array of strings:
```python
>>> s, t = T.flatten([['a', 'b'], ['c', 'd']])
>>> ft.unflatten((s, str(t)))
Segmentation fault: 11
```
In general it seems like the fasttypes codes assumes that rectangular arrays will only hold fixed-width numerical data, so that they can be unflattened as numpy arrays. But we don't have such a restriction in labrad.

There's also an issue with how fasttypes deals with nested python arrays:
```python
>>> s, t = T.flatten([[1,2],[3,4]]); (s, str(t))
('\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00\x04',
 '*2i')

>>> ft.flatten([[1,2],[3,4]])
('\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00\x04',
 '**i')

>>> ft.flatten([[1,2],[3,4]], '*2i') # even if I specify rectangular
('\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00\x04',
 '**i')
```
`labrad.types` always tries to flatten nested lists as rectangular n-dimensional arrays. If as it traverses it discovers that things are not rectangular, it bails:
```python
>>> s, t = T.flatten([[1,2],[3,4, 5]]); (s, str(t))
...
Exception: List is not rectangular.
```
In fact, the type `**i` was not even meant to be allowed; this came from the fact that in labview, you cannot have ragged arrays. If you want to do that, you have to make an array of clusters with arrays inside, and we took that over into the python implementation.
```python
>>> s, t = T.flatten([([1,2],),([3,4, 5],)]); (s, str(t))
('\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00\x03\x00\x00\x00\x04\x00\x00\x00\x05',
 '*(*i)')
```
So the fact that we parse `**i` as a valid tag was not intended; in fact we never produce such a type when flattening. Now, given that we aren't really using or supporting labview anymore, I'd be happy to drop that restriction and allow these sort of ragged arrays, but we'd need to make sure they work everywhere and also provide a way to enforce rectangularity even when flattening nested lists if that is what is desired.

ejeffrey commented:
I have pushed some unit tests for the issues @maffoo raised.

afowler commented:
Thanks Matthew, Evan, fasttypes now rejects nonrectangular nested lists and never creates type tags with multiple sequential *s.

ejeffrey commented:
Great.  I added more tests for the ragged lists and also some tests for datetime objects which are marked "expected to fail"

ejeffrey commented:
Also added tests for unflattening:

Unflattening None somehow returns a function pointer
Unflattening doesn't always raise errors on strings and arrays if the data buffer is too short

afowler commented:
Thanks Matthew and Evan for the 64-bit, short-buffer, and datetime help.
Recently pushed version passes all tests.

Regards,
Austin.

On Fri, May 8, 2015 at 10:59 PM, ejeffrey <notifications@github.com> wrote:

> Also added tests for unflattening:
>
> Unflattening None somehow returns a function pointer
> Unflattening doesn't always raise errors on strings and arrays if the data
> buffer is too short
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88#issuecomment-100435193>
> .
>
